diff --git a/pack_img.sh b/pack_img.sh
index 0331d46..e6cd0f0 100755
--- a/pack_img.sh
+++ b/pack_img.sh
@@ -2108,5 +2108,6 @@ if [ "x${PACK_TAR_IMAGE}" = "xtar_image" ]; then
 	do_pack_partition_file ${ROOT_DIR}/image/sys_partition.fex
 fi
 
-[ -e ${PACK_TOPDIR}/scripts/.hooks/post-pack ] &&
+if [ -e ${PACK_TOPDIR}/scripts/.hooks/post-pack ]; then
 	source ${PACK_TOPDIR}/scripts/.hooks/post-pack
+fi
